hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MethodProbe.java

```diff
@@ -93,61 +173,68 @@ static <S> MethodProbe createMethodProbe(Method method, Probe probe, SourceMetad
                     method.getDeclaringClass().getName(), method.getName()));
         }
 
-        if (isDouble(type)) {
+        if (type.getMapsTo() == double.class) {
             return new DoubleMethodProbe<S>(method, probe, type, sourceMetadata);
-        } else {
+        } else if (type.getMapsTo() == long.class) {
             return new LongMethodProbe<S>(method, probe, type, sourceMetadata);
+        } else {
+            throw new IllegalArgumentException(type.toString());
         }
     }
 
     static class LongMethodProbe<S> extends MethodProbe implements LongProbeFunction<S> {
-
-        LongMethodProbe(Method method, Probe probe, int type, SourceMetadata sourceMetadata) {
+        LongMethodProbe(final Method method, final Probe probe, final ProbeType type, final SourceMetadata sourceMetadata) {
             super(method, probe, type, sourceMetadata);
         }
 
+        @SuppressWarnings("checkstyle:CyclomaticComplexity")
         @Override
-        public long get(S source) throws Exception {
+        public long get(final S source) throws Throwable {
             switch (type) {
-                case TYPE_PRIMITIVE_LONG:
-                    return ((Number) method.invoke(source, EMPTY_ARGS)).longValue();
+                case TYPE_LONG_PRIMITIVE:
```

Comment:
The optimization should be applied to all fields of type byte, short, int, long. And that is why the type system needs to be enriched so that it includes all these primitives.

Reply:
> The optimization should be applied to all fields of type byte, short, int, long. And that is why the type system needs to be enriched so that it includes all these primitives.

I'm not sure that this is required, but I don't think it's in scope for this issue - do you want to raise it separately?

> perhaps the confusion is caused by the fact that this is the LongMethodProbe; but this probe can be used for fields of type byte, short, int and long.

I hadn't considered this - I've added `com.hazelcast.internal.metrics.impl.ProbeUtilsTest.main(String[])` to print all of the types a `Probe` has been found against.

Reply:
It in in scope of this issue. You want to remove the litter.

Reply:
What is needed here is something like this:

```
  @Override
        public long get(S source) throws Exception {
            switch (type) {
                case TYPE_PRIMITIVE_LONG:
                    return (long) METHOD_HANDLE.invokeExact(source);
                case TYPE_PRIMITIVE_INT:
                    return (int) METHOD_HANDLE.invokeExact(source);
                case TYPE_PRIMITIVE_BYTE:
                    return (byte) METHOD_HANDLE.invokeExact(source);
                case TYPE_PRIMITIVE_CHAR:
                    return (char) METHOD_HANDLE.invokeExact(source);
                case TYPE_LONG_NUMBER:
                    Number longNumber = (Number) method.invoke(source, EMPTY_ARGS);
                    return longNumber == null ? 0 : longNumber.longValue();
                case TYPE_MAP:
                    Map<?, ?> map = (Map<?, ?>) method.invoke(source, EMPTY_ARGS);
                    return map == null ? 0 : map.size();
                case TYPE_COLLECTION:
                    Collection<?> collection = (Collection<?>) method.invoke(source, EMPTY_ARGS);
                    return collection == null ? 0 : collection.size();
                case TYPE_COUNTER:
                    Counter counter = (Counter) method.invoke(source, EMPTY_ARGS);
                    return counter == null ? 0 : counter.get();
                case TYPE_SEMAPHORE:
                    Semaphore semaphore = (Semaphore) method.invoke(source, EMPTY_ARGS);
                    return semaphore == null ? 0 : semaphore.availablePermits();
                default:
                    throw new IllegalStateException("Unrecognized type:" + type);
            }```

Reply:
> It in in scope of this issue. You want to remove the litter.

Do you have a benchmark showing this is required? From the `MethodHandleMismatchBenchmark` you sent me, I don't think this is required when using `java.lang.invoke.MethodType.changeReturnType(Class<?>)`.

I get no litter with this.